crates/nu-parser/src/parse_literals.rs

```diff
@@ -0,0 +1,573 @@
+use crate::{lex::lex, parser::garbage, TokenContents};
+use itertools::Itertools;
+use log::trace;
+use nu_protocol::{ast::*, engine::StateWorkingSet, ParseError, Span, Spanned, Type};
+use std::{borrow::Cow, num::ParseIntError, str};
+
+/// Strip underscores for numeric literals
+fn strip_underscores(token: &[u8]) -> Cow<[u8]> {
+    if let Some(first_underscore) = token.iter().position(|b| *b == b'_') {
+        // We assume that we parse numbers (with reasonable len) and only shortly hold the
+        // allocation, thus allocate the same size for the owned case.
+        let mut owned = Vec::with_capacity(token.len());
+        owned.extend_from_slice(&token[..first_underscore]);
+        owned.extend(token[first_underscore + 1..].iter().filter(|b| **b != b'_'));
+
+        Cow::Owned(owned)
+    } else {
+        Cow::Borrowed(token)
+    }
+}
+
+/// Failed to reinterpet bytes as an ASCII string slice
+#[derive(Debug, PartialEq)]
+struct AsciiError;
+
+/// Alternative to `str::from_utf8` if you only care about ASCII characters
+///
+/// This also allows to shortcircuit following checks if you only expect ASCII characters.
+fn parse_ascii_as_str(maybe_ascii: &[u8]) -> Result<&str, AsciiError> {
+    if maybe_ascii.is_ascii() {
+        // Safe as all ASCII characters are valid UTF-8
+        Ok(unsafe { std::str::from_utf8_unchecked(maybe_ascii) })
+    } else {
+        Err(AsciiError)
+    }
+}
+
+pub fn parse_int(working_set: &mut StateWorkingSet, span: Span) -> Expression {
+    let token = working_set.get_span_contents(span);
+
+    fn parse_int_with_base(token: &[u8], span: Span, radix: u32) -> Result<Expression, ParseError> {
+        let token = strip_underscores(token);
+        if let Ok(token) = parse_ascii_as_str(&token) {
+            if let Ok(num) = i64::from_str_radix(token, radix) {
+                return Ok(Expression {
+                    expr: Expr::Int(num),
+                    span,
+                    ty: Type::Int,
+                    custom_completion: None,
+                });
+            }
+        }
+        Err(ParseError::InvalidLiteral(
+            format!("invalid digits for radix {}", radix),
+            "int".into(),
+            span,
+        ))
+    }
+
+    if token.is_empty() {
+        working_set.error(ParseError::Expected("int", span));
+        return garbage(span);
+    }
+
+    let res = if let Some(num) = token.strip_prefix(b"0b") {
+        parse_int_with_base(num, span, 2)
+    } else if let Some(num) = token.strip_prefix(b"0o") {
+        parse_int_with_base(num, span, 8)
+    } else if let Some(num) = token.strip_prefix(b"0x") {
+        parse_int_with_base(num, span, 16)
+    } else {
+        let token = strip_underscores(token);
```

Comment:
is there a reason to not just use `parse_int_with_base(..., 10)`? just the error message?

Reply:
I actually didn't know that `impl FromStr for i64` in fact just goes down to `from_str_radix(x, 10)`, would have guessed that they have a special impl for that.
https://doc.rust-lang.org/src/core/num/mod.rs.html#1400-1411